packages/core/src/internal/motion/utils.ts

```diff
@@ -0,0 +1,372 @@
+/*
+ * Copyright (c) 2016-2021 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+import {
+  AnimatableElement,
+  AnimationStatus,
+  CLARITY_MOTION_ENTER_LEAVE_PROPERTY,
+  CLARITY_MOTION_FALLBACK_DURATION_IN_MS,
+  CLARITY_MOTION_FALLBACK_EASING,
+  CLARITY_MOTION_REVERSE_ANIMATION_LABEL,
+  CLARITY_MOTION_REVERSE_ANIMATION_SUFFIX,
+  PropertyDrivenAnimation,
+  TargetedAnimation,
+  TargetedAnimationAsPropertyTuple,
+} from './interfaces.js';
+import { LogService } from '../services/log.service.js';
+import { ClarityMotion } from './motion.service.js';
+import clone from 'ramda/es/clone.js';
+import { getCssPropertyValue, isCssPropertyName } from '../utils/css.js';
+import { isPrefixedBy, isSuffixedBy, getNumericValueFromCssSecondsStyleValue, removePrefix } from '../utils/string.js';
+import { queryChildFromLightOrShadowDom } from '../utils/dom.js';
+import { allPropertiesPass, getMillisecondsFromSeconds } from '../utils/identity.js';
+
+/**
+ * runPropertyAnimations() is a utility function called by the @animate(). It is a single point of entry
+ * to execute the property-driven animations found in components like the accordion and overlay.
+ *
+ * Other utilities here are exported to break up the unit tests covering the animation/motion
+ * library. While they could be called or used in isolation, it is not recommended for most use cases.
+ *
+ * @internal
+ */
+export async function runPropertyAnimations(props: Map<string, any>, hostEl: AnimatableElement) {
```

Comment:
This function drives a family of utilities that does all the work for the animations. It amounts to a bunch of transformations on the animation config objects and finals down to a list of promises.

You can follow the comments for a pretty good guide as to what state the transformation is in at each major milestones.

Reply:
Right now, the utils have really good code coverage. One of the benefits of a more functional approach is the ability to test things from the inside out. You can test the functions that drive meta-functions like `runPropertyAnimations()` in isolation. This wasn't possible in the angular because we took an OOP approach, meaning if we had utility functions they were often private or unavailable outside of the angular component.

This meant we often had to test a component's functionality from the outside in – we had to test the results, as opposed to the pieces that did the actual work.

Another benefit of this approach is that we will have a clearer picture of what is failing if a test breaks. Testing from the outside in, you may see a test fail but have no idea what the cause of that failure might be.

Reply:
Sounds good. fwiw I was focused on the control flow logic of `runPropertyAnimations` and the things it was doing, not the calls to the delegated functions. e.g if line 86 was modified by accident were removed by accident and the call on line happened incorrectly or didn't happen when it needs to how would we know?

Reply:
oh i see. that is a valid concern. the limiting factor is how poorly karma allows us to spy on just plain functions. methods on objects? great! a standalone function coming from a module? 😱 

the easiest way to unit test that control flow is to either... break it out into its own function(s) that returns our promises and animations so we can inspect them. or build all of this into some kind of static class so we can spy on the methods. could/should the utilities be part of a general service? either `ClarityMotion` or another service to keep `ClarityMotion` light?

these are good questions i don't have good answers to. thoughts @hippee-lee and @coryrylan ?

Reply:
What about leaving a comment one future test enhancement? 

I didn't block because its working for me and I understand it. Adding tests will provide a verification check that it always does what I think it does but I'm not sure how hard that is at this point. Since you are comfortable I didn't want to block on my hang up but It is worth calling out that there could be a few places in the function logic that are hard to track down b/c they aren't tested , yet.

imo - not a blocker but worth leaving a note or a TODO for future devs.

Reply:
I would lean towards smaller functions over a class that has to be spied on. Because the test coverage is high I wont block on this either. I added some comments of possible ideas of ways to break the function up into smaller functions. Each comment block looked like a possible opportunity for extracting to a function. Even if they are not exported for tests it might make it easier to read/follow.